### ISSUE TITLE
proposal: add Dummy opcode gadget in evm circuit to enable testing more complex bytecode trace

### ISSUE THREAD

**Issue by lispc** [privacy-scaling-explorations/zkevm-circuits#440]
Add DummpyGadget  like <https://github.com/scroll-tech/zkevm-circuits/blob/63ceea081021e3e28430185f637bf90fa315bea8/zkevm-circuits/src/evm_circuit/execution/dummy.rs> 

So we can add lines like below in zkevm-circuits/src/evm_circuit/execution.rs to "mute" these opcodes, 

```
sha3_gadget: DummpyGadget<F, { ExecutionState::SHA3 }>,
```

so we can test more complex real world trace, not blocking by opcodes not implemented.

**Comment by han0110** [comment]
Having a `DummyGadget` sounds good to me. Integration test also needs EVM circuit to be provable.

**Comment by barryWhiteHat** [comment]
@lispc should i assign you to this issue?

**Comment by lispc** [comment]
> @lispc should i assign you to this issue?

That is ok. It is already implemented in my branch. Just need to pick it in a new pr

**Comment by ed255** [comment]
This sounds really good! We already have dummy implementations for opcodes in the circuit input builder.  Maybe you already thought about this, but I suggest adding a `log::warn!` when using such a dummy gadget (in the advice assign function for example) so that we can see in integration tests which opcodes we are skipping.  Something like this: https://github.com/appliedzkp/zkevm-circuits/blob/e806164a84585edc74b06cf7f476725fdb49cb33/bus-mapping/src/evm/opcodes.rs#L233

**Comment by barryWhiteHat** [comment]
I assigned you just so we can make sure to close this one once we get PR merged :)

**Comment by lispc** [comment]
No problem i will submit a PR after #402 and #335 are merged (since related codes in my branch is a bit messy and i have to clean them)

**Comment by ed255** [comment]
> No problem i will submit a PR after #376 is merged (since related codes in my branch is a bit messy and i have to clean them)

TxLogs in bus-mapping is now merged :)  So I guess this task is now unlocked?

**Comment by lispc** [comment]
yes. I currently have two implementations 

(1) the one i mentioned above, so we can add `sha3_gadget: DummyGadget<F, 2, 1, { ExecutionState::SHA3 }>,` to give it a dummy implementation. Using this method, even the state circuit can pass.

(2) add ExecutionState::DUMMY. In witness.rs, convert all unimplemented opcodes to this state. In execution.rs, just skip DUMMY and place no constraints. Using this method, every opcode can be "dummy" proved in evm circuit.

Any ideas?

**Comment by ed255** [comment]
> yes. I currently have two implementations
> 
> (1) [the one i mentioned above](https://github.com/scroll-tech/zkevm-circuits/blob/292610fe4fe4bb8df4f2a702a27729d4ff2ffa47/zkevm-circuits/src/evm_circuit/execution/dummy.rs), so we can add `sha3_gadget: DummyGadget<F, 2, 1, { ExecutionState::SHA3 }>,` to give it a dummy implementation. Using this method, even the state circuit can pass.
> 
> (2) [add ExecutionState::DUMMY.](https://github.com/scroll-tech/zkevm-circuits/pull/162/files#diff-191c4ac110f85f0e7c99e4387fabeb9041751b7aacae9f8b11dc7bcc1ae9893cR883) In witness.rs, convert all unimplemented opcodes to this state. In execution.rs, just skip DUMMY and place no constraints. Using this method, every opcode can be "dummy" proved in evm circuit.
> 
> Any ideas?

I personally prefer option 1 because it handles the push and pops from the stack, and this allows us to test the integration with the state circuit as you mention :)  Also we can use the `DummyGadget` for other execution states as needed as a placeholder until we have a proper implementation.

**Comment by lispc** [comment]
OK btw as far as i can remember adding sha3 dummy like this will help integration tests pass with oz erc20 multi transfer

**Comment by ed255** [comment]
Resolved via https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/563
